### PR TITLE
E-ink tablet displays copied recipes from ebooks and SD-cards

### PR DESCRIPTION
**Pull request by Hymore246**
#### Summary
Interface "E-ink tablet displays copied recipes from ebooks and SD-cards"

#### Purpose of change

Recipes stored in an e-ink tablet are included in the item description.

#### Describe the solution

After creating #50579, I decided to give the same treatment to the e-ink tablet. Previously, it was very hard to view recipes copied from SD cards and after #49408 added the ability store books in the tablet, it has become even more important to make this change. 

When you examine an e-ink tablet, it will display two addition sections that show what recipes have been copied into it. One section is for SD card recipes and the other is a combination of all recipes from every ebook in the tablet. These follow the same format as in #50579. Furthermore, the ebook section will only display a recipe once even if it is found in multiple books to save space and make it easier to read. This doesn't include recipes scanned from SD cards so there can be duplicates between the two sections. This is intentional since there is no other way to view SD card recipes at this time. Lastly, I added some sorting code for SD card recipes so they will be displayed in alphabetical order instead of the order they were added to the tablet.

#### Describe alternatives you've considered

Due to the fact that recipes are stored in a comma separated string (SD card recipes) and in multiple string vectors (ebook recipes) I am concerned lag might become a problem. This will probably not be an issue unless the number of recipes gets extremely high (several thousand) but is worth keeping an eye on for potential problems.

#### Testing

1. Spawn one or more e-ink tablets (and batteries).
2. Spawn several SD cards.
3. Spawn any book containing recipes.
4. Scan the book into the tablet and examine the tablet afterwards.
5. Scan SD cards until you download a recipe and examine the tablet afterwards.

#### Additional context
Example of a full e-ink tablet description containing 5 books and a few SD recipes:
eink-display1

eink-display2

### CODE REVIEW

**Comment by dseguin** [comment]
Just tested this PR. It's a cool feature!
e-ink_book

**Comment by actual-nh** [comment]
Suggestion for a later E-ink tablet (or camera?) PR: Copying maps, such as from the card(s) in #51155.

File: src/item.cpp [hunk 1]
```diff
@@ -3620,6 +3622,201 @@ void item::tool_info( std::vector<iteminfo> &info, const iteminfo_query *parts,
...
+
+            insert_separation_line( info );
+            info.emplace_back( iteminfo( "DESCRIPTION", recipe_line ) );
+
+            if( !known_recipe_list.empty() ) {
+                std::vector<std::string> sorted_known_recipes = known_recipe_list;
+                std::sort( sorted_known_recipes.begin(), sorted_known_recipes.end(), localized_compare );
+                std::string recipe_line =
+                    string_format( ngettext( "\nYou already known %1$d recipe:\n%2$s",
+                                             "\nYou already known %1$d recipes:\n%2$s",
```
Comment (Night-Pryanik) [inline src/item.cpp:3682]:
```suggestion
                    string_format( ngettext( "\nYou already know %1$d recipe:\n%2$s",
                                             "\nYou already know %1$d recipes:\n%2$s",
```

File: src/item.cpp [hunk 2]
```diff
@@ -3620,6 +3622,201 @@ void item::tool_info( std::vector<iteminfo> &info, const iteminfo_query *parts,
...
+            insert_separation_line( info );
+            info.emplace_back( iteminfo( "DESCRIPTION", recipe_line ) );
+            info.emplace_back( iteminfo( "DESCRIPTION", source_line ) );
+
+            if( !known_recipe_list.empty() ) {
+                std::vector<std::string> sorted_known_recipes = known_recipe_list;
+                std::sort( sorted_known_recipes.begin(), sorted_known_recipes.end(), localized_compare );
+                std::string recipe_line =
+                    string_format( ngettext( "\nYou already known %1$d recipe:\n%2$s",
+                                             "\nYou already known %1$d recipes:\n%2$s",
```
Comment (Night-Pryanik) [inline src/item.cpp:3787]:
```suggestion
                    string_format( ngettext( "\nYou already know %1$d recipe:\n%2$s",
                                             "\nYou already know %1$d recipes:\n%2$s",
```

**Comment by Entity-y** [comment]
Would there be a way to move recipes and books from a tablet back onto an SD card to move data between devices like another tablet, laptop, or smartphone? 

Perhaps this feature could be expanded to add new actions to the device, contained on SD cards as mentioned in #51064

**Comment by Hymore246** [comment]
> 
> 
> Would there be a way to move recipes and books from a tablet back onto an SD card to move data between devices like another tablet, laptop, or smartphone?
> 
> Perhaps this feature could be expanded to add new actions to the device, contained on SD cards as mentioned in #51064

Not at this time. It would take a fair amount of work to allow the transfer and SD card's code would need a bunch of updates to allow it to receive recipes. Smart applications are a whole other story since those would need to be designed and added individually.

**Pull request merged**